solver/app/v2/types.go

```diff
@@ -40,63 +47,82 @@ func (id OrderID) String() string {
 	return strconv.FormatUint(id.Uint64(), 10)
 }
 
-// FillInstruction returns the order fill instruction. SolverNet v1 ERC7683 orders only have one fill instruction.
-func (o Order) FillInstruction() bindings.IERC7683FillInstruction {
-	return o.Resolved.FillInstructions[0]
-}
+func newOrder(resolved OrderResolved, state OrderState, history OrderHistory) (Order, error) {
+	if err := validateResolved(resolved); err != nil {
+		return Order{}, errors.Wrap(err, "validate resolved")
+	}
 
-// DestinationSettler returns the destination settler address.
-func (o Order) DestinationSettler() (common.Address, error) {
-	settler := o.FillInstruction().DestinationSettler
-	addr, err := cast.EthAddress(settler[:])
+	fillInstr := resolved.FillInstructions[0]
+	o := Order{
+		FillInstruction:    fillInstr,
+		SourceChainID:      resolved.OriginChainId.Uint64(),
+		DestinationChainID: fillInstr.DestinationChainId,
+		Status:             state.Status,
+		AcceptedBy:         state.AcceptedBy,
+		History:            history,
+	}
+
+	settlerBz := fillInstr.DestinationSettler
+	settler, err := cast.EthAddress(settlerBz[:])
 	if err != nil {
-		return common.Address{}, errors.Wrap(err, "cast to eth addr")
+		return Order{}, errors.Wrap(err, "cast destination settler")
 	}
+	o.DestinationSettler = settler
 
-	return addr, nil
+	return o, nil
+}
+
+func validateResolved(o OrderResolved) error {
+	if o.OrderId == [32]byte{} {
+		return errors.New("order ID is empty")
+	}
+
+	if o.OriginChainId == nil {
+		return errors.New("origin chain ID is nil")
+	}
+
+	if o.FillInstructions == nil {
+		return errors.New("fill instructions are nil")
+	}
+
+	if o.MaxSpent == nil {
+		return errors.New("max spent is nil")
+	}
+
+	if len(o.FillInstructions) != 1 {
+		return errors.New("expected exactly one fill instruction")
+	}
+
+	return nil
 }
 
 // LogAttrs returns a map of order attributes for logging. It tries to extract fil origin data with known format.
 func (o Order) LogAttrs(ctx context.Context) []any {
 	attrs := []any{
 		"id", o.ID.String(),
 		"status", o.Status,
-		"accepted_by", o.AcceptedBy.Hex(),
-		"src_chain_id", o.SourceChainID(),
-		"dst_chain_id", o.DestinationChainID(),
+		"src_chain_id", o.SourceChainID,
+		"dst_chain_id", o.DestinationChainID,
 	}
 
-	fill, err := parseFillOriginData(o.FillInstruction().OriginData)
+	fill, err := parseFillOriginData(o.FillOriginData)
 	if err != nil {
 		log.Warn(ctx, "Failed to parse fill origin data", err, attrs...)
```

Comment:
logging shouldn't happen in deep inside helper functions. Logging should only happen in main logic loops. Helper functions should return errors rather.